oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/ModelName.java

```diff
@@ -30,10 +43,73 @@ public static String build(Downsampling downsampling, String modelName) {
+        List<String> indexNameList = new ArrayList<>();
+
+        List<String> whiteIndexList = new ArrayList<String>() {
+            {
+                add(EndpointInventory.INDEX_NAME);
+                add(NetworkAddressInventory.INDEX_NAME);
+                add(ServiceInventory.INDEX_NAME);
+                add(ServiceInstanceInventory.INDEX_NAME);
+            }
+        };
+
+        if (whiteIndexList.contains(modelName)) {
+            return new String[] { modelName };
+        }
+
+        DateTime startDT = DurationUtils.INSTANCE.startTimeBucket2DateTime(downsampling, startTB);
+        DateTime endDT = DurationUtils.INSTANCE.endTimeBucket2DateTime(downsampling, endTB);
+        if (endDT.isAfter(startDT)) {
+            switch (downsampling) {
+                case Month:
+                    while (endDT.isAfter(startDT)) {
+                        String indexName = build(downsampling, modelName) + "-" + YYYYMM.print(startDT);
+                        indexNameList.add(indexName);
+                        startDT = startDT.plusMonths(1);
+                    }
+                    break;
+                case Day:
+                    while (endDT.isAfter(startDT)) {
+                        String indexName = build(downsampling, modelName) + "-" + YYYYMMDD.print(startDT);
+                        indexNameList.add(indexName);
+                        startDT = startDT.plusDays(1);
+                    }
+                    break;
+                case Hour:
+                    //current hour index is also suffix with YYYYMMDD
+                    while (endDT.isAfter(startDT)) {
+                        String indexName = build(downsampling, modelName) + "-" + YYYYMMDD.print(startDT);
+                        indexNameList.add(indexName);
+                        startDT = startDT.plusDays(1);
+                    }
+                    break;
+                case Minute:
+                    //current minute index is also suffix with YYYYMMDD
+                    while (endDT.isAfter(startDT)) {
+                        String indexName = build(downsampling, modelName) + "-" + YYYYMMDD.print(startDT);
+                        indexNameList.add(indexName);
+                        startDT = startDT.plusDays(1);
+                    }
+                    break;
+                case Second:
+                    //current second index is also suffix with YYYYMMDD
+                    while (endDT.isAfter(startDT)) {
+                        String indexName = build(downsampling, modelName) + "-" + YYYYMMDD.print(startDT);
+                        indexNameList.add(indexName);
+                        startDT = startDT.plusDays(1);
+                    }
+                    break;
+                default:
+                    indexNameList.add(modelName);
+            }
+        }
+        return indexNameList.toArray(new String[0]);
```

Comment:
You covert to array at the first place, but then you covert to List again in `#filterNotExistIndex`. Seems unnecessary?

Reply:
If select a date with a long span, which dates are not in the middle, or the system has been stopped for a few days, there may be no index data for several days. If you query date  contain those days directly, searching the index will appear `index not found exception`. so, Did a filter that the date index does not exist

Reply:
Did you answer the question in the wrong context? I am not saying `filterNotExistIndex` is useless.

But based on the things you talked, I think you should establish a cache to `filterNotExistIndex` and cache the result for at least a day, considering the index could only be deleted day by day, right? Set up a guava cache with `#expireAfterWrite`?

Reply:
> You covert to array at the first place, but then you covert to List again in `#filterNotExistIndex`. Seems unnecessary?
Oh, indeed i will fix it. Convert at last minitue.

> But based on the things you talked, I think you should establish a cache to `filterNotExistIndex` and cache the result for at least a day, considering the index could only be deleted day by day, right? Set up a guava cache with `#expireAfterWrite`?
Good advice, i will try.